app/services/foreman/plugin.rb

```diff
@@ -234,15 +238,34 @@ def permission(name, hash, options = {})
 
     # Add a new role if it doesn't exist
     def role(name, permissions)
-      @default_roles[name] = permissions
+      default_roles[name] = permissions
       return false if pending_migrations || Rails.env.test?
 
       Role.transaction do
         role = Role.where(:name => name).first_or_create
         role.add_permissions!(permissions) if role.permissions.empty?
```

Comment:
Now when roles are locked and you added support for in memory searching, we don't need to add new filters only if there are no assigned already. Without the if condition it works as expected, if the permission is missing, it's added. No duplicates seems to be created. The reason why we avoided this in past was that user could set some filter condition and we'd add the permission to such filter but this is no longer possible for shipped (locked) roles.

TL;DR pls remove the if condition on this line :-)